### DISCUSSION TITLE
No Installation opam

### DISCUSSION THREAD

**Discussion by musikfast** [savonet/liquidsoap#2425] [Installation]
I cannot install Liquidsoap via opam. When I installed it, there is a lot in the Liquidsoap folder, but the service does not start at all. How can I get the service to work? The version with apt-get doesn't work either.

**Comment by caf2022** [comment]
If you are using Ubuntu 20.x+ run installation in this sequence:

Install opam:
```
add-apt-repository ppa:avsm/ppa -y
apt update -y
apt install opam -y

opam init --disable-sandboxing -a --bare
opam switch create 4.14.0
eval $(opam env --switch=4.14.0)
```
Install FFMPEG
```
cd /root
wget https://downloads.sourceforge.net/lame/lame-3.99.5.tar.gz
tar -zxf lame-3.99.5.tar.gz
cd /root/lame-3.99.5
./configure --enable-shared --enable-nasm
make
make install
make distclean

echo '/usr/local/lib' >> /etc/ld.so.conf
echo '/usr/lib' >> /etc/ld.so.conf

cd /root
git clone http://git.videolan.org/git/x264.git
cd x264
./configure --enable-shared --enable-pic --disable-asm && make && make install

wget https://ffmpeg.org/releases/ffmpeg-4.4.tar.gz
tar -zxvf ffmpeg-4.4.tar.gz
cd ffmpeg-4.4/
./configure --enable-nonfree --enable-openssl --disable-yasm --enable-libmp3lame --enable-libx264 --enable-pic --enable-pic --enable-gpl --enable-shared --enable-decoder=aac --enable-filter=aformat --enable-filter=volume --enable-filter=aresample
make
make install
ldconfig
```
Install Liquidsoap 2.0.5
```
sudo apt install frei0r-plugins-dev ladspa-sdk libavcodec-dev libavdevice-dev libavfilter-dev libavformat-dev libavutil-dev libcurl4-gnutls-dev libfaad-dev libfdk-aac-dev autoconf automake g++ libflac-dev libmad0-dev libmp3lame-dev libogg-dev libopus-dev libpcre3-dev libsamplerate0-dev libspeex-dev libtag1-dev libtheora-dev libvorbis-dev zlib1g-dev libflac-dev libmad0-dev libmp3lame-dev libogg-dev libopus-dev libpcre3-dev libsamplerate0-dev libspeex-dev libtag1-dev libtheora-dev libvorbis-dev -y

wget https://github.com/savonet/liquidsoap-release-assets/releases/download/v2.0.5/liquidsoap-2.0.5.tar.bz2
tar -xvf liquidsoap-2.0.5.tar.bz2
cd liquidsoap-2.0.5/
opam pin add liquidsoap . --no-action -y

opam install liquidsoap ocurl taglib fdkaac mad lame vorbis cry faad flac opus vorbis theora speex ogg xmlplaylist duppy ladspa dtools samplerate ffmpeg ffmpeg-avutil ffmpeg-avcodec ffmpeg-avdevice ffmpeg-av ffmpeg-avfilter ffmpeg-swresample ffmpeg-swscale frei0r -y --assume-depexts

eval $(opam env)

liquidsoap --version 
```
If some error with faad occurs, run this(reinstall instead of install):

`opam reinstall liquidsoap ocurl taglib fdkaac mad lame vorbis cry faad flac opus vorbis theora speex ogg xmlplaylist duppy ladspa dtools samplerate ffmpeg ffmpeg-avutil ffmpeg-avcodec ffmpeg-avdevice ffmpeg-av ffmpeg-avfilter ffmpeg-swresample ffmpeg-swscale frei0r -y --assume-depexts`


Hope this helps you.

**Reply by musikfast** [reply]
Unfortunately that didn't help much.
I went through all the steps but there is no entry in
/etc/liquidsoap/
/etc/init.d/Liquidsoap
and so I can't start the service.

**Comment by musikfast** [comment]
Many Thanks,

but it seems that I can't start the service. But I found a service: main-liquidsoap, which doesn't work

Is there a way to find out how to get it to run?

With the old version 1.4.4 I was able to find out more using the command: systemctl status liquidsoap, now Linux can no longer find this service.

**Reply by caf2022** [reply]
The liquidsoap was not installed as a service, you need to run it manually:

`liquidsoap /home/config.liq`

**Reply by musikfast** [reply]
in German I get this error message

Der Befehl 'liquidsoap' wurde nicht gefunden, kann aber installiert werden mit:
apt install liquidsoap


via google translator it says:
The 'liquidsoap' command was not found but can be installed with:

apt install liquidsoap

**Reply by musikfast** [reply]
I can understand that I would have to adapt the .liq file because of the forced update from 1.4.4 to 2.0.5. I'm also currently working on the Liquidsoap book, but it would be nice if liquidsoap ran at all.

**Reply by musikfast** [reply]
I might need Liquidsoap-daemon?

**Comment by toots** [comment]
Hi,

How did you install `liquidsoap`? There are several methods and the final binary might be in a different location based on that.

**Reply by musikfast** [reply]
Yes, I installed liquidsoap as described above

**Reply by musikfast** [reply]
I had the 1.4 version via apt-get install for about 1.5 years.
Now there was a problem and I wanted to install the new one. But right now I don't have a radio anymore.

**Reply by toots** [reply]
Can you please be more specific about the installation? More details help to give more precise help:
* What OS are you running?
* What installation method have you used? `apt-get` (with what package)? `opam`? 
* How do you usually run your script?

**Reply by musikfast** [reply]
Ubuntu 20.04.4 LTS
Install method `opam`
my script doesn't run

**Reply by musikfast** [reply]
I installed it as caf2022 (see above) wrote me.
Then I tried to write a script to start `liquidsoap`manually, but it doesn't work. Script and error messages :

> [Unit]
Description=LiquidSoap Stream
After=network-online.target
[Service]
ExecStart=/etc/liquidsoap/radio-run.liq
Restart=always
WorkingDirectory=/home/benjamin/.opam/4.14.0/lib
User=mod
Group=radiouser
[Install]
WantedBy=network-online.target

Error Message from `systemctl status liquidsoap:`

> ● liquidsoap.service - LiquidSoap Stream
     Loaded: loaded (/etc/systemd/system/liquidsoap.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Mon 2022-05-30 17:27:55 CEST; 1h 29min ago
    Process: 1329 ExecStart=/etc/liquidsoap/radio-run.liq (code=exited, status=203/EXEC)
   Main PID: 1329 (code=exited, status=203/EXEC)
Mai 30 17:27:54 CMLV-BF systemd[1]: liquidsoap.service: Main process exited, code=exited, status=203/EXEC
Mai 30 17:27:54 CMLV-BF systemd[1]: liquidsoap.service: Failed with result 'exit-code'.
Mai 30 17:27:55 CMLV-BF systemd[1]: liquidsoap.service: Scheduled restart job, restart counter is at 5.
Mai 30 17:27:55 CMLV-BF systemd[1]: Stopped LiquidSoap Stream.
Mai 30 17:27:55 CMLV-BF systemd[1]: liquidsoap.service: Start request repeated too quickly.
Mai 30 17:27:55 CMLV-BF systemd[1]: liquidsoap.service: Failed with result 'exit-code'.
Mai 30 17:27:55 CMLV-BF systemd[1]: Failed to start LiquidSoap Stream.

**Reply by musikfast** [reply]
FYI, I added the file in` /etc/liquidsoap/myradio.liq` manually. This folder did not exist after installation.
I just renamed the file, didn't change the content.
Now I have the following error message:

> ● liquidsoap.service - LiquidSoap Stream
     Loaded: loaded (/etc/systemd/system/liquidsoap.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Mon 2022-05-30 19:18:07 CEST; 539ms ago
    Process: 49378 ExecStart=/etc/liquidsoap/myradio.liq (code=exited, status=203/EXEC)
   Main PID: 49378 (code=exited, status=203/EXEC)
Mai 30 19:18:07 CMLV-BF systemd[1]: liquidsoap.service: Scheduled restart job, restart counter is at 5.
Mai 30 19:18:07 CMLV-BF systemd[1]: Stopped LiquidSoap Stream.
Mai 30 19:18:07 CMLV-BF systemd[1]: liquidsoap.service: Start request repeated too quickly.
Mai 30 19:18:07 CMLV-BF systemd[1]: liquidsoap.service: Failed with result 'exit-code'.
Mai 30 19:18:07 CMLV-BF systemd[1]: Failed to start LiquidSoap Stream.

**Reply by musikfast** [reply]
This is my old script with the controls for the streams that used to be in `/etc/liquidsoap/radio.liq` for version 1.4
```
# Log einrichten
set("log.file.path","/tmp/basic-radio.log")

#def crossfade(a,b)
 # def add_transition_length(_) =
  #  [("liq_transition_length","15.")]
 # end
#
 # transition =
 #   add(normalize=false,
 #         [ sequence([ blank(duration=5.),
#                      fade.in(duration=10.,b) ]),
#            fade.out(duration=10.,a) ])

#   Playlisten

playlist = playlist.safe("/home/mod/Playlist/play.txt")
emergency = playlist.safe("/home/mod/Playlist/emerg.txt")
stille = single("/home/mod/Musik/Jingle/stille.mp3")

#   Einspeisung der Sendungen:

rw = input.http("http://www.domain.de:8000/live")
radio = input.http("http://www.domain.de:8000/gllt")
vbnm = input.http("http://www.domain.de:8000/inv")

#  Als Fallback, für einen Ausfall nun security definieren radio = fallback(track_sensitive = false,
#  [radio, stille])


standard = fallback(track_sensitive=false,[radio, playlist, emergency, stille])
rwi = fallback(track_sensitive=false,[rw, radio, playlist, emergency, stille])
nmark = fallback(track_sensitive=false,[vbnm, radio, playlist, emergency, stille])

#Automatische Kompression und Normalisation

output.icecast(%mp3,
host = "domain.de",
port = 8000,
name = "Radio",
genre = "Christlich",
description = "O Land, Land, Land, höre des Herrn Wort!",
url = "https://www.my-domain.de",
user = "source",
password = "XXXXXXXX",
mount = "gll",
standard)


output.icecast(%mp3,
host = "domain.de",
port = 8000,
name = "EChB",
genre = "Christlich",
description = "Gottesdienst-Ausstrahlung",
url = "http://www.otherdomain.de",
user = "source",
password = "XXXXXXXX",
mount = "godi",
rwi)

output.icecast(%mp3,
host = "domain.de",
port = 8000,
name = "ECB",
genre = "Christlich",
description = "Gottesdienst-Ausstrahlung",
user = "source",
password = "XXXXXXXX",
mount = "vburg",
nmark)

```

**Reply by musikfast** [reply]
I repeated the installation and tried to install everything clean, with no manual script to control the service. Now nothing works again. The folders in `/etc/...` will not be created. Something doesn't fit with `opam `or is my Ubuntu defective

**Reply by musikfast** [reply]
I also tried to install version 2.0.4 - same problem

**Comment by toots** [comment]
There are a couple of considerations to understand and solve your issue:

Fundamentally, installing liquidsoap means installing the binary interpreter and its standard library. This is what the `opam` package provides. You can check by issuing a couple of commands:
```
# See where the binary is installed:
% which liquisoap
# Get a summary of options:
% liquidsoap --help\
# Get the documentation an operator:
% liquidsoap -h playlist
```
The way you run your script can be vastly different, specially nowadays with virtualization such as `docker` and `kubernetes`. 

Running as a daemon is another popular option. The `debian` package traditionally provided support for running as a daemon in the context of a debian (and perhaps debian derived like ubuntu) distribution.

Unfortunately, the debian packages shipped with our own releases do not appear to include that support anymore. We do provide daemonization scripts via https://github.com/savonet/liquidsoap-daemon.

The problem with providing support for daemonization is that there are a lot of different methods for each different distribution and we have limited resources. From our perspective, we are happy if we can provide a way to install a functional interpreter binary and its standard library to as many distributions as possible. The rest is a on best effort basis.

 I see. that you made some progress running the script as a daemon. Upgrading from `1.4.4` to `2.0.5` also requires updating your script, which could be a reason that your new daemonized scripts aren't working.

I would suggest to:
* First, run your new script manually in a console and confirm that it does work.
* Then, work on daemoizing that working script.

Lastly, as explained in the release notes, we always suggest to have a staging environment to test your script before switching/upgrading to a new version, in order to avoid downtime and unforeseen changes in behavior when upgrading your production version.

**Reply by musikfast** [reply]
> Fundamentally, installing liquidsoap means installing the binary interpreter and its standard library. This is what the `opam` package provides. You can check by issuing a couple of commands:
> 
> ```
> # See where the binary is installed:
> % which liquisoap
> # Get a summary of options:
> % liquidsoap --help\
> # Get the documentation an operator:
> % liquidsoap -h playlist
> ```

I tested it, but I don't get any indication where the binary has been installed.
I have Ubuntu 20.04.4

I suspect that the installation of the binary interpreter fails for me. Since I unfortunately have little profound knowledge of Linux and have been looking for a solution for a long time, I am writing here in the hope that someone can help me. I wrote my script for configuring the stream myself and it worked. I think I'll be able to do it in the 2.0.5 version when the binary interpreter installation finally works.
Are there any log files where I can uncover the installation errors? Maybe there are no rights for the folders to install? Shouldn't an error message appear immediately after the installation?
It always showed to me that the installation worked great and I got congratulations

**Reply by toots** [reply]
Have you tried the packages we provide for ubuntu focal (20.04)?
